packages/table/src/Table.tsx

```diff
@@ -0,0 +1,196 @@
+import React, { useReducer, ReactElement } from 'react';
+import { isComponentType } from '@leafygreen-ui/lib';
+import { css } from '@leafygreen-ui/emotion';
+import HeaderRow, { HeaderRowProps } from './HeaderRow';
+import TableHeader, { TableHeaderProps } from './TableHeader';
+import CheckboxCell from './CheckboxCell';
+import { State, Types, TableProvider, reducer } from './table-context';
+
+// * Clean TS
+// * Add nodeRef to Row
+
+const tableStyles = css`
+  border-collapse: collapse;
+  box-sizing: border-box;
+`;
+
+interface TableProps extends React.ComponentPropsWithoutRef<'table'> {
+  data?: Array<any>;
+  columns?:
+    | Array<ReactElement<HeaderRowProps>>
+    | Array<ReactElement<TableHeaderProps> | string>;
+  selectable?: boolean;
+}
+
+export default function Table({
+  columns = [],
+  data = [],
+  selectable: selectableProp = false,
+  children,
+  ...rest
+}: TableProps) {
+  const initialState: State = {
+    sort: {
+      direction: 'asc',
+    },
+    data,
+    selectable: selectableProp,
+    headerCheckState: false,
+    headerIndeterminate: false,
+  };
+
+  const [state, dispatch] = useReducer(reducer, initialState);
+
+  React.useEffect(() => {
+    const rowCheckedState: {
+      [n in number]: undefined | boolean;
+    } = {};
+
+    rows.forEach((datum, index) => {
+      if (datum.props.disabled) {
+        return;
+      }
+      rowCheckedState[index] = undefined;
+    });
+
+    if (rowCheckedState) {
+      dispatch({
+        type: Types.SetRowCheckedState,
+        payload: rowCheckedState,
+      });
+    }
+  }, [data]);
+
+  React.useEffect(() => {
+    if (state.rowCheckedState) {
+      const boolArray = Object.values(state.rowCheckedState);
+      const checkSame = boolArray.every(val => val === boolArray[0]);
+
+      if (checkSame) {
+        dispatch({
+          type: Types.ToggleHeaderIndeterminate,
+          payload: false,
+        });
+
+        if (boolArray[0]) {
+          dispatch({
+            type: Types.ToggleHeaderCheckedState,
+            payload: true,
+          });
+        } else {
+          dispatch({
+            type: Types.ToggleHeaderCheckedState,
+            payload: false,
+          });
+        }
+      } else {
+        dispatch({
+          type: Types.ToggleHeaderIndeterminate,
+          payload: true,
+        });
+      }
+    }
+  }, [state.data, state.rowCheckedState]);
+
+  let usingHeaderRow = React.useMemo(() => false, [children]);
```

Comment:
Can we move this to the first line of renderHeader? Why is it out here? What does this memo get us?

Reply:
👍